src/llama-sampling.cpp

```diff
@@ -1565,6 +1565,292 @@ struct llama_sampler * llama_sampler_init_penalties(
+                    }
+
+                    // Ensure we don't already have a duplicate matching tokenization
+                    auto its = token_sequences.equal_range(token_id);
+                    bool found = false;
+                    for (auto it = its.first; it != its.second; ++it) {
+                        if (tokenization == it->second) {
+                            found = true;
+                            break;
+                        }
+                    }
+                    if (!found) {
+                        token_sequences.emplace(token_id, tokenization);
+                    }
+                }
+            }
+        }
+    }
+}
+
+
+
+static const char * llama_sampler_dry_name(const struct llama_sampler * /*smpl*/) {
+    return "dry";
+}
+
+static void llama_sampler_dry_accept(struct llama_sampler * smpl, llama_token token) {
+    auto * ctx = (llama_sampler_dry *) smpl->ctx;
+    if (ctx->dry_penalty_last_n == 0) {
+        return;
+    }
+
+    ctx->last_tokens.push_back(token);
+}
+
+static void llama_sampler_dry_apply(struct llama_sampler * smpl, llama_token_data_array * cur_p) {
+    auto * ctx = (llama_sampler_dry *) smpl->ctx;
+
+    if (ctx->dry_multiplier == 0.0f || ctx->dry_base == 0.0f) {
+        return;
+    }
+
+    int32_t effective_dry_penalty_last_n = (ctx->dry_penalty_last_n == -1) ? ctx->total_context_size : std::max(ctx->dry_penalty_last_n, 0);
+    int last_n_repeat = std::min(std::min((int)ctx->last_tokens.size(), effective_dry_penalty_last_n), ctx->total_context_size);
+
+    if (last_n_repeat <= ctx->dry_allowed_length) {
+        return;
+    }
+
+    ctx->dry_repeat_count.assign(last_n_repeat, 0);
+    ctx->dry_max_token_repeat.clear();
+
+    // Step 1: Look for restart sequences
+    int rep_limit = last_n_repeat;
+    for (int i = 0; i < last_n_repeat; ++i) {
+        llama_token token = ctx->last_tokens.rat(i);
+        auto its = ctx->dry_processed_breakers.equal_range(token);
+        if (its.first == ctx->dry_processed_breakers.end()) {
+            continue;
+        }
+        int longest_match = -1;
+        for (auto it = its.first; it != its.second; ++it) {
+            int seq_len = (int)it->second.size();
+            if (seq_len > longest_match && seq_len <= (int)i) {
+                bool match = true;
+                for (int offset = 0; offset < seq_len; ++offset) {
+                    if (it->second[offset] != ctx->last_tokens.rat(i - offset - 1)) {
+                        match = false;
+                        break;
+                    }
+                }
+                if (match) {
+                    longest_match = seq_len;
+                }
+            }
+        }
+        if (longest_match >= 0) {
+            rep_limit = i - longest_match;
+            break;
+        }
+    }
+    if (rep_limit <= ctx->dry_allowed_length) {
+        return;
+    }
+
+    // Step 2: Z-algorithm implementation
```

Comment:
> If it’s any help the [koboldcpp version](https://github.com/LostRuins/koboldcpp/blob/c38d1ecc8d8f4bcfe2f959584002ddf7a80b8e9b/gpttype_adapter.cpp#L743) of this has a lot of extra comments that go more in-depth on the intricacies of the Z-algorithm and the clever way they implement it. Most of the core implementation here is exactly the same except for the bits I had to change due to the ring_buffer accessing from the end.

Oh yes this is very useful. Then it might be relevant to link to where the implementation comes from in a comment near the code. Not sure if `kobold.cpp` being under AGPLv3 will cause problems, though.

(cc @pi6am because they wrote <https://github.com/LostRuins/koboldcpp/pull/982>)

Reply:
Oh… yikes. 😬

I feel very naive now for not looking deeper into the licensing ramifications there. My apologies.

I doubt that @LostRuins would want to relicense just this one portion of his code, which is of course totally understandable, so my assumption is barring that, this very likely forces out this critical part of the PR. 

While I really do admire @pi6am’s implementation, I think the best move at this point is to try returning to @l3utterfly’s and my older version of the code as a base (so no Z-algorithm, which I know @p-e-w prefers to not use anyway, so there’s that at least!). 

If I can get that plugged in and working well I will try to rewrite the overlapping sequence breaker code myself so that we are in the clear from a licensing perspective.

Anyway, thanks for all the feedback @compilade, I will focus on this issue first since it obviously impacts such a critical part of the code.

Reply:
The copyright holder of that portion of Kobold's code is the author of the implementation, @pi6am. Not the maintainer of the project (unless the two signed a copyright assignment agreement). So you would need their permission to relicense that code, which they might be willing to give.

Personally, I think that the MIT license used by llama.cpp is a terrible choice for projects like this, and this issue is one of the reasons. Licensing under MIT or other permissive licenses basically means you're gifting your code to businesses to put behind closed gates, while at the same time limiting what code you yourself can reuse from other open source projects. It's the worst of both worlds.

Reply:
@p-e-w Well, that is fascinating, thank you for clarifying that. I am learning a lot through this whole process!

So to follow up on this idea, I went ahead and just posted a message on the original Koboldcpp PR page pinging @pi6am in the hopes that a notification through there might work better than one here.

Reply:
Yup you'd have to get permission from @pi6am as they are the author. I personally have no issues.

Reply:
yea MIT is a bad choice in this cruel modern world. 
Unless we assume LM studio or x other million profit business actually alllocate 10% of their annual budget back to foss projects that they used. But it's all too late unless ggerna bro and other mantainer decide to change it.

Reply:
But assuming that even if we get both algorithm to work,you would still need a proper
-cli flag 
-documentation(or else no one know there were 2 algorithm actually implemented.lol. Precious effort undiscovered by thousands of people)
-frontend that properly implement to utilize it like(sillytavern implement it,etc,or else someone that use frontend may not have it enabled,UNLESS llama.cpp can *force* using a certain algorithm without needing any input from frontend)

All complete to fully benefit everyone.

Reply:
I don't have any objections to porting any or all of my DRY implementation from koboldcpp into llama.cpp. I officially give permission to incorporate it into llama.cpp under the MIT license.

That said, it would be nice to credit the koboldcpp implementation in comments. I also think including some of the other comments from my implementation would be worthwhile to make the code easier to understand.

In particular, I didn't write the Z-algorithm implementation from scratch; I ported a public-domain Python implementation from https://github.com/ivanyu/string-algorithms/blob/master/z_algorithm.py which I credited in a comment. It's not necessary to provide credit to this based on its license, but I think it's polite to do so.

Reply:
> I don't have any objections to porting any or all of my DRY implementation from koboldcpp into llama.cpp. I officially give permission to incorporate it into llama.cpp under the MIT license.

That is kind of you and very appreciated @pi6am, thank you! I had come up with a workable solution that implemented multi-token sequence breakers mixed with the non-z-algorithm approach that I was going to try to implement if your code was not available, but I will be the first to admit that it was nowhere near as optimized as your code, so I really appreciate having your blessing and being able to use your implementation.

> That said, it would be nice to credit the koboldcpp implementation in comments. I also think including some of the other comments from my implementation would be worthwhile to make the code easier to understand.
> 
> In particular, I didn't write the Z-algorithm implementation from scratch; I ported a public-domain Python implementation from https://github.com/ivanyu/string-algorithms/blob/master/z_algorithm.py which I credited in a comment. It's not necessary to provide credit to this based on its license, but I think it's polite to do so.

Agreed, and done on all counts in the commit I just made. Thanks again!